src/sentry/services/hybrid_cloud/filter_query.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import abc
+from typing import TYPE_CHECKING, Any, Callable, Generic, List, Optional, Protocol, TypeVar
+
+from django.db.models import QuerySet
+
+if TYPE_CHECKING:
+    from sentry.api.serializers import Serializer
+    from sentry.services.hybrid_cloud.auth import AuthenticationContext
+    from sentry.services.hybrid_cloud.user import APIUser
+
+
+class AsSerializer(Protocol):
+    def as_serializer(self) -> Serializer:
+        pass
+
+
+FILTER_ARGS = TypeVar("FILTER_ARGS")  # A typedict
+RPC_RESPONSE = TypeVar("RPC_RESPONSE")
+SERIALIZER_ENUM = TypeVar("SERIALIZER_ENUM", bound=AsSerializer)
+BASE_MODEL = TypeVar("BASE_MODEL")
+
+# In the future, this ought to be a pass through type that does not get double serializer, and which cannot be
+# inspected by code.
+OpaqueSerializedResponse = Any
+
+
+# A collection of classes that can be used to quickly provision interfaces & implementations for basic RPC Hybrid Cloud service implementations.
+# It enforces several niceties:
+#   * Type safety at call sites: the public API can be inferred, unknown arguments are rejected,
+#     all filter arguments are optional, and nullable filters are explicitly Optional[...]
+#   * Filter argument validation: an arbitrary function can be used to ensure that proper filters are applied
+#     e.g. we don't want to just query for every row, so we can enforce that at least one filter key is provided
+#   * A standard interface across most of our simple rpc services
+
+# This class can be inherited from to add this functionality to the public interface for a service.
+# E.g. class UserService(FilterQueryInterface[
+#        UserQueryArgs, APIUser, Union[UserSerializerResponse, UserSerializerResponseSelf]
+#      ], InterfaceWithLifecycle):
+#         ...
+class FilterQueryInterface(Generic[FILTER_ARGS, RPC_RESPONSE, SERIALIZER_ENUM], abc.ABC):
+    @abc.abstractmethod
+    def serialize_many(
+        self,
+        *,
+        filter: FILTER_ARGS,
+        as_user: Optional[APIUser] = None,
+        auth_context: Optional[AuthenticationContext] = None,
+        # An API Serializer instance to render results
+        serializer: Optional[Serializer] = None,
+    ) -> List[Any]:
+        pass
+
+    @abc.abstractmethod
+    def get_many(self, *, filter: FILTER_ARGS) -> List[RPC_RESPONSE]:
+        pass
+
+
+class FilterQueryDatabaseImpl(
```

Comment:
Double whammy is that the generics in the impl class here is not the same as the abstract one.

You have to really know that 1.  you must use both of these classes   2.  you can need type both correctly   3.  you need to fit things in the tholes that are consistent with other usages, but it's not at all clear.

I don't know how to improve upon it in general.  I think we're just facing the limitations of python typing.

Reply:
Also, I don't feel comfortable having, for instance, the `BASE_MODEL` be part of the type generics of the base.  A major, major motivation of the `__init__.py` and `impl.py` split was specifically because managing model imports in the hybrid cloud services is intractably frustrating over time.

Reply:
Are we trying to apply an abstraction too early or applying one where there isn't enough of common pattern? I think adopting the `filter` parameter with a typed dict is a good convention, perhaps the serializer and model types don't need to be a formal generic and we can use an implicit convention where each method defines the types it uses?

Reply:
> the BASE_MODEL be part of the type generics of the base

It isn't, just a part of the database implementation

Reply:
> Are we trying to apply an abstraction too early

There are enough examples for us to know we want something that looks like this. That said, I do think the filter+TypeDict combo is a lot of the value and very little of the overhead. how do you feel about that compromise for now @corps?